dataRetriever/resolvers/headerResolver.go

```diff
@@ -100,12 +107,18 @@ func (hdrRes *HeaderResolver) resolveHeaderFromNonce(key []byte) ([]byte, error)
 		return nil, dataRetriever.ErrInvalidNonceByteSlice
 	}
 
-	//Step 2. search the nonce-key pair
+	//Step 2. search the nonce-key pair first in datapool, second in storage
 	value, _ := hdrRes.hdrNonces.Get(nonce)
 	hash, ok := value.([]byte)
-
 	if hash == nil || !ok {
-		return nil, nil
+		hash, err = hdrRes.hdrNoncesStorage.Get(key)
+		if err != nil {
+			log.Debug(err.Error())
+		}
+
+		if len(hash) == 0 {
+			return nil, nil
```

Comment:
This kind of situations should be resolved generally. We need to trust that if err is nil then the object is not nil and vice versa. Please add a task in Jira for this problem to be discussed and to be resolved generally

Reply:
Ok, user story EN-2581 has been created

Reply:
I think there's no sense in using the output "hash" if there was an error reported. 
We could just return nil, err. We should also avoid returning nil, nil, instead we can return nil, ErrNotFound or something similar.

Reply:
I think we should discuss about this as using an error defined in implementation package will render the decoupling through interfaces useless by tying the higher level package to one of the implementing package. The higher level package will have something like
```GO
hash, err := implementingPackage.Get(...)
if err != nil{
     if err == implementingPackage.ErrKeyNotFound{
           //do something special here
           //return something
     }
     return nil, err
}
The problem worsens when we have more implementations. Having a "core" kind of package that will hold common errors doesn't look too appealing for me.

The idea coined with Camil is to return something in this only special case where there is no error (cacher, storer, trie, etc. perfomed ok) but the data is simply not there. We can implement a kind of "empty", not nil objects for this purpose only. In the case of []byte we could return make([]byte, 0) and in the case of interface{} we could return struct{}{}. In this way we keep our programming pattern in which we can not have situations like (nil, nil) or (object, error) and also we can easily check that the returned object is an empty one, telling that the implementation used couldn't retrieve the required data because it does not exists (not that something malfunctioned). We also need to rework the get func from level db.

```